stream_alert/alert_processor/outputs.py

```diff
@@ -199,6 +196,245 @@ def dispatch(self, **kwargs):
+        """
+        return {'api': 'https://api.pagerduty.com'}
+
+    def get_user_defined_properties(self):
+        """Get properties that must be asssigned by the user when configuring a new PagerDuty
+        event output. This should be sensitive or unique information for this use-case that
+        needs to come from the user.
+
+        Every output should return a dict that contains a 'descriptor' with a description of the
+        integration being configured.
+
+        PagerDuty also requires a routing_key that represents this integration. This
+        value should be masked during input and is a credential requirement.
+
+        Returns:
+            OrderedDict: Contains various OutputProperty items
+        """
+        return OrderedDict([
+            ('descriptor',
+             OutputProperty(description='a short and unique descriptor for this '
+                                        'PagerDuty integration')),
+            ('token',
+             OutputProperty(description='the token for this PagerDuty integration',
+                            mask_input=True,
+                            cred_requirement=True)),
+            ('service_key',
+             OutputProperty(description='the service key for this PagerDuty integration',
+                            mask_input=True,
+                            cred_requirement=True)),
+            ('escalation_policy',
+             OutputProperty(description='the name of the default escalation policy'))
+        ])
+
+    @staticmethod
+    def _get_endpoint(base_url, endpoint):
+        """Helper to get the full url for a PagerDuty Incidents endpoint.
+
+        Args:
+            base_url (str): Base URL for the API
+            endpoint (str): Endpoint that we want the full URL for
+
+        Returns:
+            str: Full URL of the provided endpoint
+        """
+        return os.path.join(base_url, endpoint)
+
+    def _check_exists_get_id(self, filter_str, url, target_key):
+        """Generic method to run a search in the PagerDuty REST API and return the id
+        of the first occurence from the results.
+
+        Args:
+            filter_str (str): The query filter to search for in the API
+            url (str): The url to send the requests to in the API
+            target_key (str): The key to extract in the returned results
+
+        Returns:
+            str: ID of the targeted element that matches the provided filter or
+                 False if a matching element does not exists.
+        """
+        params = {
+            'query': '"{}"'.format(filter_str)
+        }
+        resp = self._get_request(url, params, self.headers, False)
+        if not self._check_http_response(resp):
+            return False
+
+        response = resp.json()
+
+        # If there are results, get the first occurence from the list
+        target = response.get(target_key, False)
+        if response and target:
```

Comment:
Suggestion: returning early if there is ever the opportunity to do so, instead of having to do compound `if` checks throughout a function helps with decrease confusion and increase readability. For instance, this could be make more readable via:

```
response = resp.json()
if not response:
    return False           # Maybe even add a LOGGER.error(...) here (?)

# If there are results, get the first occurence from the list
target = response.get(target_key, False)
if not target:
    return False           # Maybe add a LOGGER.error(...) here too (?)

return target[0]['id']
```

Theoretically the last if/return can be consolidated into just:
`return response['target_key'][0]['id'] if 'target_key' in response else False`


Just food for thought :)

Reply:
I like replacing the last if/else with that super pythonic statement